core/common/src/files/FileSystem.kt

```diff
@@ -0,0 +1,147 @@
+     * @throws kotlinx.io.IOException if the [path] could not be deleted.
+     */
+    public fun delete(path: Path, mustExist: Boolean = true)
+
+    /**
+     * Creates a directory tree represented by the [path].
+     * If [path] already exists then the method throws [kotlinx.io.IOException] when [mustCreate] is `true`.
+     * The call will attempt to create only missing directories.
+     * The method is not atomic and if it fails after creating some
+     * directories these directories will not be deleted automatically.
+     * Permissions for created directories are platform-specific.
+     *
+     * @param path the path to be created.
+     * @param mustCreate the flag indicating that existence of [path] should be treated as an error,
+     * by default it is `false`.
+     *
+     * @throws kotlinx.io.IOException when [path] already exists and [mustCreate] is `true`.
+     * @throws kotlinx.io.IOException when the creation of one of the directories fails.
+     */
+    public fun createDirectories(path: Path, mustCreate: Boolean = false)
+
+    /**
+     * Atomically renames [source] to [destination] overriding [destination] if it already exists.
+     *
+     * When the file system does not support atomic move of [source] and [destination] corresponds to different
+     * file systems and the operation could not be performed atomically,
+     * [UnsupportedOperationException] is thrown.
+     *
+     * @param source the path to rename.
+     * @param destination desired path name.
+     *
+     * @throws kotlinx.io.files.FileNotFoundException when the [source] does not exist.
+     * @throws kotlinx.io.IOException when the move failed.
+     * @throws kotlin.UnsupportedOperationException when the file system does not support atomic move.
+     */
+    public fun atomicMove(source: Path, destination: Path)
+
+    /**
+     * Returns [Source] to read from a file represented by the [path].
+     *
+     * How a source will read the data is implementation-specific and failures caused
+     * by the missing file or, for example, lack of permissions may not be reported immediately,
+     * but postponed until the source will try to fetch data.
+     *
+     * @param path the path to read from.
+     *
+     * @throws kotlinx.io.files.FileNotFoundException when the file does not exist.
+     * @throws kotlinx.io.IOException when it's not possible to open the file for reading.
+     */
+    public fun read(path: Path): Source
+
+    /**
+     * Returns [Sink] to write into a file represented by the [path].
+     * File will be created if it does not exist yet.
+     *
+     * How a sink will write the data is implementation-specific and failures caused,
+     * for example, by the lack of permissions may not be reported immediately,
+     * but postponed until the sink will try to store data.
+     *
+     * @throws kotlinx.io.IOException when it's not possible to open the file for writing.
+     */
+    public fun write(path: Path): Sink
```

Comment:
I have a question regarding API around `read` and `write`.
If to look at Okio they have 2 API:
* `source(Path)` and `sink(Path)` - to just obtain `Source/Sink` and use it somewhere later (result is **not** buffered source/sink)
* `read(Path)` and `write(Path)` - to use `Source/Sink` directly in place writing or reading data (lambda provides buffered source/sink

Here, we have some mix of both naming and functionality and I found current kotlinx-io way a bit confusing because while method is called `read` we don't actually perform reading yet but just `provide source for reading`. In this case naming methods `source/sink` instead of `read/write` shows intent on what we want to do better.

Also I don't really know, but may be it will be better to return `Raw*` variants here, instead of already buffered.
But that's also more a question on conventions/defaults, not only kotlinx-io but also other dependent libraries, should provide.
Should `RawSource/RawSink` should be just mostly implementation detail and buffered variants should be provided to user, or better to expose `Raw*` variants?

Reply:
I agree that `sink/source` are better name in this context, thanks!

And it's better to return `Raw*` variants as they could be then wrapped into other decorators (to compress/decompress data, for example) and having a buffered sink/source will only add an overhead

Reply:
Done